gputils/doc/gplib.1
.TH gplib 1 "(c) 2002, 2003 James Bowman, Craig Franklin"
.SH NAME
gplib - GNU PIC Librarian
.SH SYNOPSIS
.B gplib [options] library [member]
.SH WARNING
The information in this man page is an extract from the full documentation of
gputils and is limited to the meaning of the options.  For complete and 
current documentation, refer to "gputils.ps" located in the gputils docs 
directory.
.SH DESCRIPTION
.B gplib
creates, modifies, and extracts from COFF archives for Microchip (TM) PIC 
(TM) micro-controllers.
.B gplib
is part of gputils.  Check the gputils(1) manpage for details on other GNU 
PIC utilities.
.SH OPTIONS
.TP
.B -c, --create
Create a new library.
.TP
.B -d, --delete
Delete member from library.
.TP
.B -h, --help
Show the usage message. 
.TP
.B -n, --no-index
Don't add symbol index.
.TP
.B -q, --quiet
Quiet mode.
.TP
.B -r, --replace
Add or replace member from library.
.TP
.B -s, --symbols
List global symbols in library.
.TP
.B -t, --list
List members in library.
.TP
.B -v, --version
Show the version.
.TP
.B -x, --extract
Extract member from library.
.SH SEE ALSO
gputils(1)
.SH AUTHOR
Craig Franklin <craigfranklin@comcast.net>
.SH COPYRIGHT
Copyright (C) 2002, 2003 James Bowman, Craig Franklin

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.


gputils/ChangeLog
2014-01-25  Molnar Karoly  <molnarkaroly@users.sourceforge.net>
  * doc/gputils.lyx, header/Makefile.am, header/Makefile.in,
    header/p12f1612.inc, header/p12lf1612.inc, header/p16f1613.inc,
    header/p16lf1613.inc, header/p16f1xxx.inc, libgputils/gpcfg-table.c,
    libgputils/gpprocessor.c, lkr/Makefile.am, lkr/Makefile.in,
    lkr/12f1612_g.lkr, lkr/16f1613_g.lkr, lkr/12lf1612_g.lkr,
    lkr/16lf1613_g.lkr:
    New devices based on the mplabx-2.0.

  * header/p12f1571.inc, header/p12f1572.inc, header/p12lf1571.inc,
    header/p12lf1572.inc, header/p16f1703.inc, header/p16f1707.inc,
    header/p16f1713.inc, header/p16f1716.inc, header/p16f1782.inc,
    header/p16f1783.inc, header/p16f1784.inc, header/p16f1786.inc,
    header/p16f1787.inc, header/p16f1788.inc, header/p16f1789.inc,
    header/p16lf1703.inc, header/p16lf1707.inc, header/p16lf1713.inc,
    header/p16lf1716.inc, header/p16lf1782.inc, header/p16lf1783.inc,
    header/p16lf1784.inc, header/p16lf1786.inc, header/p16lf1787.inc,
    header/p16lf1788.inc, header/p16lf1789.inc, lkr/*_g.lkr:
    Updated based on the mplabx-2.0.

2014-01-22  Molnar Karoly  <molnarkaroly@users.sourceforge.net>
  * header/p12f1822.inc, header/p12f1840.inc, header/p12lf1822.inc,
    header/p12lf1840.inc, header/p12lf1840t39a.inc, header/p12lf1840t48a.inc:
    Added the missing CONFIG2 Options: _DEBUG_ON, _DEBUG_OFF
    "#62 Missing CONFIG2 DEBUG file definition in some pic12f inc files"
    thanks to Ludrao.

2013-11-09  Molnar Karoly  <molnarkaroly@users.sourceforge.net>
  * doc/gputils.lyx, header/Makefile.am, header/Makefile.in,
    header/p16f1703.inc, header/p16f1705.inc, header/p16f1707.inc,
    header/p16f1709.inc, header/p16lf1703.inc, header/p16lf1705.inc,
    header/p16lf1707.inc, header/p16lf1709.inc,
    libgputils/gpcfg-table.c, libgputils/gpprocessor.c,
    lkr/16f1703_g.lkr, lkr/16f1705_g.lkr, lkr/16f1707_g.lkr,
    lkr/16f1709_g.lkr, lkr/16lf1703_g.lkr, lkr/16lf1705_g.lkr,
    lkr/16lf1707_g.lkr, lkr/16lf1709_g.lkr, 
    lkr/Makefile.am, lkr/Makefile.in:
    New devices based on the mplabx-1.95.

  * header/p12f1571.inc, header/p12f1572.inc, header/p12f1822.inc,
    header/p12lf1571.inc, header/p12lf1572.inc, header/p16f1784.inc,
    header/p16lf1784.inc,
    lkr/12f1572_g.lkr, lkr/12lf1572_g.lkr, lkr/16f1509_g.lkr,
    lkr/16f1783_g.lkr, lkr/16f1936_g.lkr, lkr/16f1937_g.lkr,
    lkr/16lf1509_g.lkr, lkr/16lf1783_g.lkr, lkr/16lf1906_g.lkr,
    lkr/16lf1907_g.lkr, lkr/16lf1936_g.lkr, lkr/16lf1937_g.lkr:
    Updated based on the mplabx-1.95.

  * AUTHORS:
    Corrected my email.

2013-10-12  Borut Razem  <borutr@users.sourceforge.net>
  * libgputils/gpmemory.c: fixed b_memory_get()

2013-10-11  Borut Razem  <borutr@users.sourceforge.net>
  * libgputils/gpmemory.c: fixed bug
    #269 18F26j50, gplink crash on WPDIS

2013-09-25  Borut Razem  <borutr@users.sourceforge.net>
  * gpasm/scan.l, gpasm/ppscan.l: fixed bug
    #270 Gputils doesn't compile with newer versions of bison and flex

2013-09-08  Molnar Karoly  <molnarkaroly@users.sourceforge.net>
  * doc/gputils.lyx, header/Makefile.am, header/Makefile.in,
    header/p12f1571.inc, header/p12f1572.inc, header/p12lf1571.inc,
    header/p12lf1572.inc, header/p16f1713.inc, header/p16f1716.inc,
    header/p16f1829lin.inc, header/p16lf1713.inc, header/p16lf1716.inc,
    libgputils/gpcfg-table.c, libgputils/gpprocessor.c,
    lkr/12f1571_g.lkr, lkr/12f1572_g.lkr, lkr/12lf1571_g.lkr,
    lkr/12lf1572_g.lkr, lkr/16f1713_g.lkr, lkr/16f1716_g.lkr,
    lkr/16f1829lin_g.lkr, lkr/16lf1713_g.lkr, lkr/16lf1716_g.lkr,
    lkr/Makefile.am, lkr/Makefile.in:
    New devices based on the mplabx-1.9.

2013-09-08  Molnar Karoly  <molnarkaroly@users.sourceforge.net>
  * header/*.inc:
    Updated based on the mplabx-1.9.

2013-08-09  Borut Razem  <borutr@users.sourceforge.net>
  * config.guess, config.sub: updated from git head
  * depcomp: updated
  * README: updated MinGW build info

2013-08-05  Borut Razem  <borutr@users.sourceforge.net>
  * gplink/gplink.c, libgputils/gpcofflink.c, libgputils/gpcofflink.h,
    gplink/gplink.h.in, doc/gputils.lyx, man/gplink.1.in,
    man/fr/gplink.1.in: implemented --mplink-compatble in gplink
  * gplink/testsuite/gplink.project/asmfiles/op14e.asm,
    gplink/testsuite/gplink.project/lkrfiles/op14e.lkr,
    gplink/testsuite/gplink.project/hexfiles/op14e.hex:
    added gplink regression tests for bug
    #268 PIC14e: HIGH directive not setting bit-7

2013-08-02  Borut Razem  <borutr@users.sourceforge.net>
  * gpasm/testsuite/gpasm.project/asmfiles/op14e.asm,
    gpasm/testsuite/gpasm.project/listfiles/op14e.lst,
    gpasm/testsuite/gpasm.project/hexfiles/op14e.hex,
    gpasm/testsuite/gpasm.project/objasm/op14e.asm,
    gpasm/testsuite/gpasm.project/objlstfiles/op14e.lst,
    gpasm/testsuite/gpasm.project/objfiles/op14e.o:
    added gpasm regression tests for bug
    #268 PIC14e: HIGH directive not setting bit-7

2013-08-01  Borut Razem  <borutr@users.sourceforge.net>
  * gpasm/evaluate.c, libgputils/gpcofflink.c, libgputils/gpprocessor.c,
    libgputils/gpprocessor.h: fixed bug
    #268 PIC14e: HIGH directive not setting bit-7
  * gputils/gpvo.c: display symbolic value of relocation type

2013-07-24  Borut Razem  <borutr@users.sourceforge.net>
  * libgputils/gpcofflink.c, libgputils/gpprocessor.c:
    applied modified patch [PATCH] Bad warning for BRA on 16F code,
    thanks to Richard Hodges

2013-06-30  Molnar Karoly  <molnarkaroly@users.sourceforge.net>
  * header/p16f753.inc, p16hv753.inc, header/p16f1704.inc,
    header/p16f1708.inc, header/p16lf1704.inc, header/p16lf1708.inc,
    lkr/16f753_g.lkr, lkr/16hv753_g.lkr:
    Fresh based on the mplabx-1.85.

2013-05-17  Borut Razem  <borutr@users.sourceforge.net>
  * gpasm/directive.c: generate error if CONFIG directive is used with
    non 16 bit devices

2013-05-14  Borut Razem  <borutr@users.sourceforge.net>
  * doc/html-help/Makefile.am, doc/html-help/Makefile.in:

gputils/gputils/gplib.c
/* GNU PIC Librarian
   Copyright (C) 2001, 2002, 2003, 2004, 2005
   Craig Franklin

    Copyright (C) 2016 Molnar Karoly <molnarkaroly@users.sf.net>

This file is part of gputils.

gputils is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2, or (at your option)
any later version.

gputils is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with gputils; see the file COPYING.  If not, write to
the Free Software Foundation, 59 Temple Place - Suite 330,
Boston, MA 02111-1307, USA.  */

#include "stdhdr.h"

#include "libgputils.h"
#include "gplib.h"

struct gplib_state state = {
  AR_NULL,              /* default mode, do nothing */
  0,                    /* number of objects */
};

static symbol_table_t *definition_tbl = NULL;
static symbol_table_t *symbol_index   = NULL;

#define GET_OPTIONS "cdhnqrstvx"

static struct option longopts[] =
{
  { "create",   no_argument, NULL, 'c' },
  { "delete",   no_argument, NULL, 'd' },
  { "extract",  no_argument, NULL, 'x' },
  { "help",     no_argument, NULL, 'h' },
  { "no-index", no_argument, NULL, 'n' },
  { "quiet",    no_argument, NULL, 'q' },
  { "replace",  no_argument, NULL, 'r' },
  { "symbols",  no_argument, NULL, 's' },
  { "list",     no_argument, NULL, 't' },
  { "version",  no_argument, NULL, 'v' },
  { NULL,       no_argument, NULL, '\0'}
};

/*------------------------------------------------------------------------------------------------*/

static void
_show_usage(void)
{
  printf("Usage: gplib [options] library [member]\n");
  printf("Options: [defaults in brackets after descriptions]\n");
  printf("  -c, --create       Create a new library.\n");
  printf("  -d, --delete       Delete member from library.\n");
  printf("  -h, --help         Show this usage message.\n");
  printf("  -n, --no-index     Don't add symbol index.\n");
  printf("  -q, --quiet        Quiet mode.\n");
  printf("  -r, --replace      Add or replace member from library.\n");
  printf("  -s, --symbols      List global symbols in library.\n");
  printf("  -t, --list         List members in library.\n");
  printf("  -v, --version      Show version.\n");
  printf("  -x, --extract      Extract member from library.\n\n");
  printf("Report bugs to:\n");
  printf("%s\n", PACKAGE_BUGREPORT);
  exit(0);
}

/*------------------------------------------------------------------------------------------------*/

static void
_select_mode(enum lib_modes mode)
{
  if (state.mode == AR_NULL) {
    state.mode = mode;
  } else {
    gp_error("multiple library operations selected");
  }
}

/*------------------------------------------------------------------------------------------------*/

/* return the object name without the path */

static char *
_object_name(char *file_name)
{
  char *name;

#ifdef HAVE_DOS_BASED_FILE_SYSTEM
  for (name = file_name + strlen(file_name) - 1; name >= file_name; --name) {
    if ((name[0] == UNIX_PATH_CHAR) || (name[0] == PATH_SEPARATOR_CHAR)) {
      return ++name;
    }
  }

  return file_name;
#else
  name = strrchr(file_name, PATH_SEPARATOR_CHAR);
  if (name != NULL) {
    return ++name;
  }
  else {
    return file_name;
  }
#endif
}

/*------------------------------------------------------------------------------------------------*/

static gp_boolean
_has_path(const char *file_name)
{
  char *name;

  name = strrchr(file_name, PATH_SEPARATOR_CHAR);
#ifdef HAVE_DOS_BASED_FILE_SYSTEM
  if (name != NULL) {
    name = strrchr(file_name, UNIX_PATH_CHAR);
  }
#endif

  return ((name == NULL) ? false : true);
}

/*------------------------------------------------------------------------------------------------*/

int main(int argc, char *argv[])
{
  int           c;
  int           i;
  gp_boolean    usage          = false;
  gp_boolean    update_archive = false;
  gp_boolean    no_index       = false;
  gp_archive_t *object         = NULL;
  gp_coff_t     type;

  gp_init();

  /* symbols are case sensitive */
  definition_tbl = sym_push_table(NULL, false);
  symbol_index   = sym_push_table(NULL, false);

  while ((c = getopt_long(argc, argv, GET_OPTIONS, longopts, NULL)) != EOF) {
    switch (c) {
    case '?':
    case 'h':
      usage = true;
      break;

    case 'c':
      _select_mode(AR_CREATE);
      break;

    case 'd':
      _select_mode(AR_DELETE);
      break;

    case 'n':
      no_index = true;
      break;

    case 'q':
      gp_quiet = true;
      break;

    case 'r':
      _select_mode(AR_REPLACE);
      break;

    case 's':
      _select_mode(AR_SYMBOLS);
      break;

    case 't':
      _select_mode(AR_LIST);
      break;

    case 'v':
      fprintf(stderr, "%s\n", GPLIB_VERSION_STRING);
      exit(0);
      break;

    case 'x':
      _select_mode(AR_EXTRACT);
      break;
    }

    if (usage) {
      break;
    }
  }

  if (optind < argc) {
    /* fetch the library name */
    state.filename = argv[optind++];
    /* some operations require object filenames or membernames */
    for ( ; optind < argc; optind++) {
      state.objectname[state.numobjects] = argv[optind];
      if (state.numobjects >= MAX_OBJ_NAMES) {
        gp_error("exceeded maximum number of object files");
        break;
      }
      state.numobjects++;
    }
  } else {
    usage = true;
  }

  /* User did not select an operation */
  if (state.mode == AR_NULL) {
    usage = true;
  }

  /* User did not provide object names */
  if ((state.mode != AR_LIST) && (state.mode != AR_SYMBOLS) && (state.numobjects == 0)) {
    usage = true;
  }

  if (usage) {
    _show_usage();
  }

  /* if we are not creating a new archive, we have to read an existing one */
  if (state.mode != AR_CREATE) {
    if (gp_identify_coff_file(state.filename) != GP_COFF_ARCHIVE) {
      gp_error("\"%s\" is not a valid archive file", state.filename);
      exit(1);
    }
    else {
      state.archive = gp_archive_read(state.filename);
    }
  }

  /* process the option */
  i = 0;
  switch (state.mode) {
  case AR_CREATE:
  case AR_REPLACE: {
    while (i < state.numobjects) {
      type = gp_identify_coff_file(state.objectname[i]);

      if ((type != GP_COFF_OBJECT_V2) && (type != GP_COFF_OBJECT)) {
        gp_error("\"%s\" is not a valid object file", state.objectname[i]);
        break;
      }
      else {
        state.archive = gp_archive_add_member(state.archive, state.objectname[i],
                                              _object_name(state.objectname[i]));
      }
      i++;
    }
    update_archive = true;
    break;
  }

  case AR_DELETE: {
    while (i < state.numobjects) {
      if (_has_path(state.objectname[i])) {
        gp_error("invalid object name \"%s\"", state.objectname[i]);
        break;
      }

      object = gp_archive_find_member(state.archive, state.objectname[i]);
      if (object == NULL) {
        gp_error("object \"%s\" not found", state.objectname[i]);
        break;
      }
      else {
        state.archive = gp_archive_delete_member(state.archive, state.objectname[i]);
      }
      i++;
    }
    update_archive = true;
    break;
  }

  case AR_EXTRACT: {
    while (i < state.numobjects) {
      if (_has_path(state.objectname[i])) {
        gp_error("invalid object name \"%s\"", state.objectname[i]);
        break;
      }

      object = gp_archive_find_member(state.archive, state.objectname[i]);
      if (object == NULL) {
        gp_error("object \"%s\" not found", state.objectname[i]);
        break;
      }
      else {
        if (!gp_archive_extract_member(state.archive, state.objectname[i])) {
          gp_error("can't write file \"%s\"", state.objectname[i]);
          exit(1);
          break;
        }
      }
      i++;
    }
    break;
  }

  case AR_LIST:
    gp_archive_list_members(state.archive);
    break;

  case AR_SYMBOLS: {
    if (gp_archive_have_index(state.archive) == 0) {
      gp_error("this archive has no symbol index");
    }
    else {
      gp_archive_read_index(symbol_index, state.archive);
      gp_archive_print_table(symbol_index);
    }
    break;
  }

  case AR_NULL:
  default:
    assert(0);
  }

  /* If the archive is being modified remove the old symbol index */
  if (update_archive) {
    state.archive = gp_archive_remove_index(state.archive);
  }

  /* check for duplicate symbols */
  gp_archive_make_index(state.archive, definition_tbl);

  /* add the symbol index to the archive */
  if (update_archive && (!no_index)) {
    state.archive = gp_archive_add_index(definition_tbl, state.archive);
  }

  /* write the new or modified archive */
  if (update_archive && (gp_num_errors == 0)) {
    if (!gp_archive_write(state.archive, state.filename)) {
      gp_error("can't write the new archive file");
      exit(1);
    }
  }

  return ((gp_num_errors > 0) ? EXIT_FAILURE : EXIT_SUCCESS);
}
